### ISSUE TITLE
GDB fails to start, VS crashes - "channel has not been opened" from shellexec

### ISSUE THREAD

**Issue by fabianoliver** [microsoft/VSLinux#367]
Hi,

I was just testing Linux support on a CMake project, and successfully compiled the application on the "remote" machine via VS. However, debugging fails. When trying to debug, I'm getting the following build output:

> Starting: "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\..\CommonExtensions\Microsoft\Linux\Linux\shellexec.exe" /s -509788251 /p 19848 /c /opt/rh/devtoolset-8/root/usr/bin/gdb --interpreter=mi --tty=/dev/pts/1
> shellexec 1.0
> Copyright (C) 2016 Microsoft Corporation. All rights reserved.
> Unhandled exception: System.InvalidOperationException: The channel has not been opened, or the open has not yet been confirmed.
>   at liblinux.Ssh.Channels.Channel.CreateRemoteChannelInfoNotAvailableException()
>   at liblinux.Ssh.Channels.Channel.get_RemoteChannelNumber()
>   at liblinux.Ssh.Channels.Channel.SendData(Byte[] data)
>   at liblinux.Ssh.ShellStream.WriteLine(String line)
>   at liblinux.Shell.StreamingShell.WriteLine(String line)
>   at ExecuteShell.shellexec.RunShell()
>   at ExecuteShell.shellexec.Run(String[] args)
>   at ExecutShell.shellexec.Main(String[] args)
> "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\..\CommonExtensions\Microsoft\Linux\Linux\shellexec.exe" exitedw ith code -532462766 (0xE0434352)

Which is followed by an error popup in VS,

> Unable to start debugging. Unable to establish a connection to GDB. The following message was written to stderr:
> 
> (... see Exception pasted above)

Visual Studio either completely freezes afterwards, or (more commonly) just crashes.

Maybe irrelevant, but also - the first time I'm trying to debug or build (eg either right click on CMakeLists -> Debug from Build Folder), the build fails, just with a plain message "Build failed."  in the CMake output window without any further details. Only trying to build/debug again will eventually "succeed", but then crash as described when trying to launch GBD.

VERY occasionally, it seems that attaching a debugger works. The debuggee process still exists, though that might be perfectly possible (at the time, I was testing a cmake output that isn't usually intended to be run by itself, so this probably just bugs out and throws) - but at least it seems like GDB connected there. However, this happened only twice in several dozens of attempts,

> Warning: Debuggee TargetArchitecture not detected, assuming x_86_64.
> The thread 0xfaa has exited with code 0 (0x0).
> The program 'name_of_my_process..exe' has exited with code 177 (0xb1)

I'm using:
Microsoft Visual Studio Enterprise 2017
VisualStudio.15.Release/15.9.8+28307.481
Visual C++ for Linux Development - 1.0.9.28218

And connecting (under the root account) to a a local VM via VirtualBox that is running on:
RHEL 6.9
OpenSSH version 7.9, compiled from portable release (OpenSSH_7.9p1, OpenSSL 1.0-1e-fips 11 Feb 2013)

**Comment by itodirel** [comment]
@paulmaybee can you take a look?

**Comment by paulmaybee** [comment]
I am not able to reproducing your behavior on RHEL 6.9 (with OpenSSH_5.3p1, OpenSSL 1.0.1e-fips 11 Feb 2013).
Are you able to keep an SSH session (Putty or otherwise) open to your VM? What version of gdb are you using (I am using 7.2-92.el6)? If you switch modes to gdbserver in your project properties does that work any better?

**Comment by fabianoliver** [comment]
Sorry for the late reply, I was travelling.

I'm using GDB 8.2 - though to me it looks like the Exception occurs before GDB is even launched in the first place.
Connecting via Putty works fine.

To make things even more odd: It just worked - but only once. VS connected, and launched GDB successfully. This was with exactly the same configuration and same steps of building/launching as any other time.
Anytime before that one successful launch, and anytime thereafter, I have been, and still am facing the original issue though.

I'm attaching three files - the SSH log + Visual Studio (shellexec) output from the attempt when it worked.
And the SSH log from an attempt where it didn't. In case of the latter, the exception in visual studio occured ca. at 4:12:43 server time.

[nocrash_serverlog.txt](https://github.com/Microsoft/VSLinux/files/3067647/nocrash_serverlog.txt)
[nocrash_shellexec_output.txt](https://github.com/Microsoft/VSLinux/files/3067648/nocrash_shellexec_output.txt)
[crash_at_4-12-43_serverlog.txt](https://github.com/Microsoft/VSLinux/files/3067649/crash_at_4-12-43_serverlog.txt)

**Comment by itodirel** [comment]
@paulmaybee any ideas?

**Comment by itodirel** [comment]
We have added logging since, please provide them when reporting additional information. We are not able to reproduce at this, and it is also possible we already addressed this issue. We have addressed many issues in shellexec, debugging and other areas of the Cross Plat Linux experience. If you still experience an issue here, please open an issue on the Developer Community.

We have retired this issue list, here on Github. We now adopted and are using the Developer Community (https://developercommunity.visualstudio.com/spaces/62/index.html), for Cross Plat and Linux experiences (and C++ generally), which we are actively monitoring. Please use the Developer Community, to report issues or suggestions for bugs or new experiences. Further discussions on this thread here on Github might be unmonitored. If there is further comments or context to this current issue, please create a new issue on the Developer Community, and provide the context there. For any existing issues, which we can or able to reproduce, or for anything that is actionable to us and understood, we will provide an update right here in this thread. Additionally, for new suggestions or new issues, which at this time are in our roadmap, or for which we believe and have evidence that broadly and/or significantly benefit everyone, we will open a suggestion ticket, and link it here. For anything else, we recommend to please open a ticket on the Developer Community. That isn't because we will not look at the problem, or not investigate it otherwise, but simply because we need additional information. And for features, we are tracking them on the Developer Community.
